lib/json/common.rb

```diff
@@ -612,16 +612,13 @@ class << self
   # Output:
   #   {"foo":[0,1],"bar":{"baz":2,"bat":3},"bam":"bad"}
   def dump(obj, anIO = nil, limit = nil, kwargs = nil)
-    if anIO and limit.nil?
-      anIO = anIO.to_io if anIO.respond_to?(:to_io)
-      unless anIO.respond_to?(:write)
-        if kwargs.nil? and anIO.is_a?(Hash)
-          kwargs = anIO
-        else
-          limit = anIO
-        end
-        anIO = nil
-      end
+    io_limit_opt = [anIO, limit, kwargs].compact
+    kwargs = io_limit_opt.pop if io_limit_opt.last.is_a?(Hash)
+    anIO, limit = io_limit_opt
+    if anIO.respond_to?(:to_io)
+      anIO = anIO.to_io
+    elsif limit.nil? && !anIO.respond_to?(:write)
+      anIO, limit = nil, anIO
```

Comment:
Fixed the inconsistency
```ruby
dump(obj, io, nil) #=> io.to_io was called
dump(obj, io, 10) #=> io.to_io was not called
```

And if `to_io` is called, the result of `to_io` should be an IO, no need to check for `respond_to?(:write)`